Examples/Algorithms/Digitization/src/HitSmearing.cpp

```diff
@@ -11,95 +11,125 @@
+  for (auto&& [moduleGeoId, moduleSimHits] : groupByModule(simHits)) {
     // check if we should create hits for this surface
-    const auto is = m_surfaces.find(moduleGeoId);
-    if (is == m_surfaces.end()) {
+    auto surfaceItr = m_surfaces.find(moduleGeoId);
+    if (surfaceItr == m_surfaces.end()) {
       continue;
     }
+    const Acts::Surface* surface = surfaceItr->second;
+
+    // use iterators manually so we can retrieve the hit index in the container
+    for (auto ih = moduleSimHits.begin(); ih != moduleSimHits.end(); ++ih) {
+      const auto& simHit = *ih;
+      const auto simHitIdx = simHits.index_of(ih);
 
-    // smear all truth hits for this module
-    const Acts::Surface* surface = is->second;
-    for (const auto& hit : moduleHits) {
       // transform global position into local coordinates
-      auto lpResult = surface->globalToLocal(ctx.geoContext, hit.position(),
-                                             hit.unitDirection(), 0.5_um);
-      Acts::Vector2D lp{0., 0.};
+      auto lpResult = surface->globalToLocal(ctx.geoContext, simHit.position(),
+                                             simHit.unitDirection(), 0.5_um);
       if (not lpResult.ok()) {
         ACTS_ERROR("Global to local transformation did not succeed.");
         return ProcessCode::ABORT;
-      } else {
-        lp = lpResult.value();
       }
 
-      // smear truth to create local measurement
-      Acts::BoundVector loc = Acts::BoundVector::Zero();
-      loc[Acts::eBoundLoc0] = lp[0] + m_cfg.sigmaLoc0 * stdNormal(rng);
-      loc[Acts::eBoundLoc1] = lp[1] + m_cfg.sigmaLoc1 * stdNormal(rng);
+      // create smeared local measurement
+      Acts::Vector2D loc = lpResult.value();
+      loc[0] += m_cfg.sigmaLoc0 * stdNormal(rng);
+      loc[1] += m_cfg.sigmaLoc1 * stdNormal(rng);
 
-      // create source link at the end of the container
-      auto it = sourceLinks.emplace_hint(sourceLinks.end(), *surface, hit, 2,
-                                         loc, cov);
-      // ensure hits and links share the same order to prevent ugly surprises
-      if (std::next(it) != sourceLinks.end()) {
-        ACTS_FATAL("The hit ordering broke. Run for your life.");
-        return ProcessCode::ABORT;
-      }
+      // the measurement container is unordered and the index under which the
+      // measurement will be stored is known before adding it.
+      Index hitIdx = measurements.size();
+      IndexSourceLink sourceLink(*surface, hitIdx);
+      ConcreteMeasurement meas(surface->getSharedPtr(), sourceLink, cov, loc);
+
+      // add to output containers. since the input is already geometry-order,
+      // new elements in geometry containers can just be appended at the end.
+      measurements.emplace_back(std::move(meas));
+      sourceLinks.emplace_hint(sourceLinks.end(), std::move(sourceLink));
+      // no hit merging -> only one mapping per digitized hit.
+      hitParticlesMap.emplace_hint(hitParticlesMap.end(), hitIdx,
```

Comment:
Why do we write out the `hitParticlesMap` here again since we already have this in the `CsvPlanarClusterReader`. Is there any difference between them?

Reply:
You can call this in a setup without the `CsvPlanarClusterReader` where only the simulated hits are available, e.g. during the simulation or once we added a `CsvSimHit{Reader,Writer}`. Logically, the digitizer is the only algorithm that knows how hits and particles are connected and is the place where this maps gets created.

Previously, this information was implicitly stored in the `SimSourceLink` and the `CsvPlaranClusterWriter` was just decoding it and wrote it too disk. With this PR, the mapping is always created explicitly.

Reply:
That makes sense. It's nice to have this direct bonding here.